Site/SiteWebApplication.php

```diff
@@ -735,6 +735,49 @@ public function relocate($uri, $secure = null, $append_sid = null,
 	}
 
 	// }}}
+	// {{{ public function relocateWithQueryString()
+
+	/**
+	 * Relocates to another URI while maintaining the current query string.
+	 *
+	 * This is useful for keeping things like google UTM strings being
+	 * maintained in automatic redirects.
+	 *
+	 * @param string $uri the URI to relocate to.
+	 * @param boolean $secure optional. Whether or not the base href should be
+	 *                         a secure URI. The default value of null
+	 *                         maintains the same security as the current page.
+	 *                         This parameter only has an effect if the
+	 *                         <i>$uri</i> is relative.
+	 * @param boolean $append_sid optional. Whether or not to append the
+	 *                             session identifier to the URI. If null, this
+	 *                             is determined automatically by the
+	 *                             {@link SiteSessionModule::appendSessionId()}
+	 *                             method.
+	 * @param boolean $permanent Whether or not to the relocate is permanent.
+	 *                            Set true for urls that are permanently
+	 *                            moved.
+	 */
+	public function relocateWithQueryString($uri, $secure = null,
+		$append_sid = null, $permanent = false)
+	{
+		$source_uri = parse_url($_SERVER['REQUEST_URI']);
```

Comment:
Use `parse_url($foo, PHP_URL_QUERY)` to not need an interstitial variable to compute the query string.